proofs/signatures/drat.plf

```diff
@@ -0,0 +1,215 @@
+                                      (cln tt)
+                                      ((clc l c') (CMap_has_bottom rest))))))
+
+; Return a new CMap with one copy of this clause removed.
+(program CMap_remove_clause ((c clause) (cs CMap)) CMap
+         (match cs
+                (CMapn (fail CMap))
+                ((CMapc i c' cs')
+                 (match (clause_eq c c')
+                        (tt cs')
+                        (ff (CMapc i c' (CMap_remove_clause c cs')))))))
+
+; Turn a cnf into a CMap, where the indices start at `start` and increment
+(program CMap_from_cnf ((f cnf) (start mpz)) CMap
+         (match f
+                (cnfn CMapn)
+                ((cnfc c f') (CMapc start c (CMap_from_cnf f' (mp_add 1 start))))))
+
+
+; return (c1 union (c2 \ ~l))
+; Significant for how a RAT is defined.
+(program clause_pseudo_resolvent ((c1 clause) (l lit) (c2 clause)) clause
+         (clause_dedup (clause_append c1
+                                      (clause_remove_all (lit_flip l) c2))))
+
+; Given a formula, `cs` and a clause `c`, return all pseudo resolvants, i.e. all
+;     (c union (c' \ ~head(c)))
+;   for c' in cs, where c' contains ~head(c)
+(program collect_pseudo_resolvents ((cs CMap) (c clause)) cnf
+         (match cs
+                (CMapn cnfn)
+                ((CMapc i c' cs')
+                 (let rest_of_resolvents (collect_pseudo_resolvents cs' c)
+                   (match (clause_contains_lit c' (lit_flip (clause_head c)))
+                          (tt (cnfc (clause_pseudo_resolvent
+                                      c
+                                      (clause_head c)
+                                      c')
+                                    rest_of_resolvents))
+                          (ff rest_of_resolvents))))))
+
+; =============================================================== ;
+; Unit Propegation implementation (manipulates global assignment) ;
+; =============================================================== ;
+; See the lrat file for a description of the global assignment.
+
+(declare UnitSearchResult type)
+; There was a unit, and this is the (previoiusly floating) literal that is now satified.
+(declare USRUnit (! l lit UnitSearchResult))
+; There was an unsat clause
+(declare USRBottom UnitSearchResult)
+; There was no unit.
+(declare USRNoUnit UnitSearchResult)
+
+; Searches through the clauses, looking for a unit clause.
+;  Returns
+;    USRBottom    if there is an unsat clause
+;    (USRUnit l)  if there is a unit, with lit l
+;    USRNoUnit    if there is no unit
+(program unit_search ((cs CMap)) UnitSearchResult
+         (match cs
+                (CMapn USRNoUnit)
+                ((CMapc i c cs')
+                 (match (clause_check_unit_and_maybe_mark c)
+                        (MRSat (unit_search cs'))
+                        ((MRUnit l) (USRUnit l))
+                        (MRUnsat USRBottom)
+                        (MRNotUnit (unit_search cs'))))))
+
+
+
```

Comment:
1. Good catch, will do!
2. Yeah, it would improve efficiency to remove now-satisfied unit clauses, as well as any satisfied clauses. I'll do that.
3. Yes, yes there are. I'll add this to the documentation.